### PR TITLE
expression: add an argument `EvalType` to method `Column.Reset` to make it safer

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/13487.
Add a new argument `EvalType` to method `Column.Reset` to let it reset `Column.elemBuf` according to this argument, which makes it safer.

Here is [a simple example](https://github.com/pingcap/tidb/pull/13488/files#diff-05f90244f392d371054b04f4a1df8d3dR981) to show why it is not safe if we don't add this argument.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: util/chunk/column.go [hunk 1]
```diff
@@ -91,8 +92,30 @@ func (c *Column) isFixed() bool {
...
+	case types.ETDuration:
+		c.ResizeGoDuration(0, false)
+	case types.ETJson:
+		c.ReserveJSON(0)
+	default:
+		panic(fmt.Sprintf("invalid EvalType %v", eType))
+	}
+}
+
+func (c *Column) reset() {
```
Comment (SunRunAway) [inline util/chunk/column.go]: Add a comment for this method.

Comment (qw4990) [reply util/chunk/column.go]: updated, PTAL

File: expression/generator/time_vec.go [hunk 2]
```diff
@@ -213,23 +213,6 @@ func (b *{{.SigName}}) vecEval{{ .Output.TypeName }}(input *chunk.Chunk, result
...
-		}
-		output := sum.String()
-	{{ else if eq .SigName "builtinAddDateAndStringSig" }}
-		{{ template "ConvertStringToDuration" . }}
-		fsp0 := int8(b.args[0].GetType().Decimal)
-		sum, err := types.Duration{Duration: arg0, Fsp: fsp0}.Add(arg1Duration)
-		if err != nil {
-			return err
-		}
-		output := sum.String()
```
Comment (SunRunAway) [inline expression/generator/time_vec.go]: Why this modification?

File: util/chunk/column.go [hunk 3]
```diff
@@ -91,8 +92,32 @@ func (c *Column) isFixed() bool {
...
+		c.ResizeGoDuration(0, false)
+	case types.ETJson:
+		c.ReserveJSON(0)
+	default:
+		panic(fmt.Sprintf("invalid EvalType %v", eType))
+	}
+}
+
+// reset resets this Column but doesn't modify the elemBuf field since
+// it doesn't know what type the data is.
```
Comment (SunRunAway) [inline util/chunk/column.go]: suggestion: reset resets the underlying data of this Column but doesn't modify its data type.

File: expression/bench_test.go [hunk 4]
```diff
@@ -785,8 +785,8 @@ func genVecExprBenchCase(ctx sessionctx.Context, funcName string, testCase vecEx
 	}
 	cols := make([]Expression, len(testCase.childrenTypes))
 	input = chunk.New(fts, 1024, 1024)
-	for i, eType := range testCase.childrenTypes {
-		fillColumn(eType, input, i, testCase)
+	for i := range testCase.childrenTypes {
+		fillColumn(fts[i].EvalType(), input, i, testCase)
```
Comment (SunRunAway) [inline expression/bench_test.go]: Could you explain why this change?
This will make the `childrenTypes` enty of `testCase` useless.

Comment (qw4990) [reply expression/bench_test.go]: updated.
